bundle/jsky.app.ot/src/main/scala/jsky/app/ot/gemini/inst/OIWFSVignettingFeature.scala

```diff
@@ -0,0 +1,185 @@
+   * @param offsetPosX   the X screen coordinate for the selected offset
+   * @param offsetPosY   the X screen coordinate for the selected offset
+   * @param translateX   translate resulting figure by this amount of pixels in X
+   * @param translateY   translate resulting figure by this amount of pixels in Y
+   * @param basePosX     the X screen coordinate for the base position
+   * @param basePosY     the Y screen coordinate for the base position
+   * @param oiwfsDefined set to true if an OIWFS position is defined (otherwise
+   */
+  override def _updateFigureList(guidePosX:  Double, guidePosY:  Double,
+                                 offsetPosX: Double, offsetPosY: Double,
+                                 translateX: Double, translateY: Double,
+                                 basePosX:   Double, basePosY:   Double,
+                                 oiwfsDefined: Boolean): Unit = {
+    _figureList.clear()
+    addOffsetConstrainedPatrolFieldAtBase(basePosX, basePosY)
+    addPatrolFieldAtBase(offsetPosX + translateX, offsetPosY + translateY)
+    if (oiwfsDefined)
+      addProbeArm(offsetPosX, offsetPosY, translateX, translateY, basePosX, basePosY)
+  }
+
+  /**
+   * Check to see if the offset intersection is empty, and if so, issue a warning.
+   *
+   * @return Some(singleton list with a warning) if the offset intersection is empty, and
+   *         None otherwise.
+   */
+  private def getMessagesAsScala: Option[java.util.Collection[TpeMessage]] = {
+    lazy val offsets = getContext.offsets.scienceOffsetsJava
+    for {
+      ctx         <- _iw.getMinimalObsContext.asScalaOpt
+      patrolField <- probe.getCorrectedPatrolField(ctx).asScalaOpt
+      if patrolField.outerLimitOffsetIntersection(offsets).isEmpty
+    } yield List(NoValidRegionWarning).asJavaCollection
+  }
+
+  override def getMessages =
+    getMessagesAsScala.asGeminiOpt
+
+  /**
+   * We always force an update due to buggy code or the effort not worth doing to maintain
+   * old offset lists.
+   */
+  override protected def _needsUpdate(inst: SPInstObsComp, tii: TpeImageInfo): Boolean = true
+}
+
+case object OIWFSVignettingFeature {
+  // Drawing specifications.
+  lazy val ProbeArmColor        = Color.RED
+  lazy val Blocked              = AlphaComposite.getInstance(AlphaComposite.SRC_OVER, 0.5F)
+
+  // A TPE warning if there is no valid region.
+  lazy val NoValidRegionWarning = TpeMessage.warningMessage("No valid OIWFS region. Check offset positions.")
+}
+
+
+
+// Concrete instances.
+object Flamingos2OIWFSFeature extends OIWFSVignettingFeature(Flamingos2OiwfsGuideProbe.instance, F2OiwfsProbeArm, "Flamingos2 OIWFS")
+object GmosNorthOIWFSFeature  extends OIWFSVignettingFeature(GmosOiwfsGuideProbe.instance, GmosNorthOiwfsProbeArm, "GMOS OIWFS")
+object GmosSouthOIWFSFeature  extends OIWFSVignettingFeature(GmosOiwfsGuideProbe.instance, GmosSouthOiwfsProbeArm, "GMOS OIWFS")
```

Comment:
I think it's really cool that we just need one TPE feature for these now.  Nice.